phd-tests/tests/src/boot_order.rs

```diff
@@ -0,0 +1,673 @@
+        DiskBackend::InMemory { readonly: true },
+        16,
+    );
+
+    // `boot-disk` is the implicitly-created boot disk made from the default guest OS artifact.
+    //
+    // explicitly boot from it later, so OVMF has to try and fail to boot `unbootable`.
+    cfg.boot_order(vec!["boot-disk"]); //, "unbootable"]);
+
+    let mut vm = ctx.spawn_vm(&cfg, None).await?;
+    vm.launch().await?;
+    vm.wait_to_boot().await?;
+
+    // If the guest doesn't have an EFI partition then there's no way for boot order preferences to
+    // be persisted.
+    let mountline = vm.run_shell_command("mount | grep /boot/efi").await?;
+
+    if !mountline.contains(" on /boot/efi type vfat") {
+        warn!("guest doesn't have an EFI partition, cannot manage boot order");
+    }
+
+    // Try adding a few new boot options, then add them to the boot order, reboot, and make sure
+    // they're all as we set them.
+    if !run_long_command(&vm, &format!("ls {}", efipath(&bootvar(0xffff))))
+        .await?
+        .is_empty()
+    {
+        warn!("guest environment already has a BootFFFF entry; is this not a fresh image?");
+    }
+
+    let boot_num: u16 = {
+        let bytes = read_efivar(&vm, BOOT_CURRENT_VAR).await?;
+        u16::from_le_bytes(bytes.try_into().unwrap())
+    };
+
+    // The entry we booted from is clearly valid, so we should be able to insert a few duplicate
+    // entries. We won't boot into them, but if something bogus happens and we did boot one of
+    // these, at least it'll work and we can detect the misbehavior.
+    //
+    // But here's a weird one: if we just append these to the end, on reboot they'll be moved
+    // somewhat up the boot order. This occurrs both if setting variables through `efibootmgr` or
+    // by writing to /sys/firmware/efi/efivars/BootOrder-* directly. As an example, say we had a
+    // boot order of "0004,0001,0003,0000" where boot options were as follows:
+    // * 0000: UiApp
+    // * 0001: PCI device 4, function 0
+    // * 0003: EFI shell (Firmware volume+file)
+    // * 0004: Ubuntu (HD partition 15, GPT formatted)
+    //
+    // If we duplicate entry 4 to new options FFF0 and FFFF, reset the boot order to
+    // "0004,0001,0003,0000,FFF0,FFFF", then reboot the VM, the boot order when it comes back
+    // up will be "0004,0001,FFF0,FFFF,0003,0000".
+    //
+    // This almost makes sense, but with other devices in the mix I've seen reorderings like
+    // `0004,0001,<PCI 16.0>,0003,0000,FFF0,FFFF` turning into
+    // `0004,0001,FFF0,FFFF,<PCI 16.0>,0003,0000`. This is particularly strange in that the new
+    // options were reordered around some other PCI device. It's not the boot order we set!
+    //
+    // So, to at least confirm we *can* modify the boot order in a stable way, make a somewhat less
+    // ambitious change: insert the duplicate boot options in the order directly after the option
+    // they are duplicates of. This seems to not get reordered.
```

Comment:
i cannot make sense of this, but it doesn't seem to be a bug resulting from anything _we're_ doing. i'd describe the overall behavior as "kind of works, and probably not something to debug in Propolis, maybe something to better understand in OVMF"..

Reply:
I think this is because of ovmf trying to reconcile the EFI boot order with what's in the qemu fwcfg - https://github.com/oxidecomputer/edk2/blob/edd72391086a9a297bf9aa011fe53b449165e556/OvmfPkg/Library/QemuBootOrderLib/QemuBootOrderLib.c#L1911

Reply:
Ah, basically what Greg had [called out](https://github.com/oxidecomputer/rfd/pull/751#discussion_r1617885908) on the RFD

Reply:
yeah exactly. the specifically confusing thing about this, to me, is that the duplicate FFF0 and FFFF entries remain in the boot order at all. i would expect them to get dropped late in `SetBootOrderFromQemu`, when it calls into `BootOrderComplete` and gets [here](https://github.com/oxidecomputer/edk2/blob/propolis/edk2-stable202105/OvmfPkg/Library/QemuBootOrderLib/QemuBootOrderLib.c#L1784-L1789).

between this and fwcfg overriding guest-configured boot order generally, i'm leaning towards actual configurations Propolis receives only naming a single boot device. if any of this changes because of a change in OVMF at some point, i want the test to fail as an indicator that Something Changed, but it feels like we shouldn't be relying on specific semantics here